java/flight/flight-sql/src/main/java/org/apache/arrow/flight/sql/FlightSQLClientUtils.java

```diff
@@ -0,0 +1,219 @@
+     * @param client The client. FlightSQLPreparedStatement does not maintain this resource.
+     * @param sql    The query.
+     */
+    public FlightSQLPreparedStatement(FlightClient client, String sql) {
+      this.client = client;
+
+      final Iterator<Result> preparedStatementResults = client.doAction(new Action("GetPreparedStatement",
+              Any.pack(FlightSQL.ActionGetPreparedStatementRequest
+                      .newBuilder()
+                      .setQuery(sql)
+                      .build())
+                      .toByteArray()));
+
+      preparedStatementResult = FlightSQLUtils.unpackAndParseOrThrow(
+              preparedStatementResults.next().getBody(),
+              ActionGetPreparedStatementResult.class);
+
+      invocationCount = 0;
+      isClosed = false;
+    }
+
+    /**
+     * Returns the Schema of the resultset.
+     *
+     * @return the Schema of the resultset.
+     */
+    public Schema getResultSetSchema() {
+      if (resultSetSchema == null && preparedStatementResult.getDatasetSchema() != null) {
+        resultSetSchema = Schema.deserialize(preparedStatementResult.getDatasetSchema().asReadOnlyByteBuffer());
+      }
+      return resultSetSchema;
+    }
+
+    /**
+     * Returns the Schema of the parameters.
+     *
+     * @return the Schema of the parameters.
+     */
+    public Schema getParameterSchema() {
+      if (parameterSchema == null && preparedStatementResult.getParameterSchema() != null) {
+        parameterSchema = Schema.deserialize(preparedStatementResult.getParameterSchema().asReadOnlyByteBuffer());
+      }
+      return parameterSchema;
+    }
+
+    /**
+     * Executes the prepared statement query on the server.
+     *
+     * @return a FlightInfo object representing the stream(s) to fetch.
+     * @throws IOException if the PreparedStatement is closed.
+     */
+    public FlightInfo executeQuery() throws IOException {
+      if (isClosed) {
+        throw new IOException("Prepared statement has already been closed on the server.");
+      }
+
+      final FlightDescriptor descriptor = FlightDescriptor
+              .command(Any.pack(CommandPreparedStatementQuery.newBuilder()
+                      .setClientExecutionHandle(
+                              ByteString.copyFrom(ByteBuffer.allocate(Long.BYTES).putLong(invocationCount++)))
+                      .setPreparedStatementHandle(preparedStatementResult.getPreparedStatementHandle())
+                      .build())
+                      .toByteArray());
+
+      return client.getInfo(descriptor);
+    }
+
+    /**
+     * Executes the prepared statement update on the server.
+     *
+     * @return the number of rows updated.
+     */
+    public long executeUpdate() {
+      throw Status.UNIMPLEMENTED.asRuntimeException();
+    }
+
+    @Override
+    public void close() {
+      isClosed = true;
+      final Iterator<Result> closePreparedStatementResults = client.doAction(new Action("ClosePreparedStatement",
```

Comment:
Multiple calls to close() should behave idempotently. We shouldn't re-issue the close request if close() was already called.